test/sqllogictest/explain/optimized_plan_as_text.slt

```diff
@@ -26,94 +29,570 @@ SELECT * FROM t WHERE a IS NOT NULL
+----
+Explained Query
+  Let
+    Project #2, #4
+      Join
+        ArrangeBy keys=[[#0]]
+          Get l0
+        ArrangeBy keys=[[#0]]
+          Union
+            Get l3
+            Map null
+              Union
+                Negate
+                  Project #0
+                    Get l3
+                Get l1
+        Union
+          Get l4
+          Map null
+            Union
+              Negate
+                Project #0
+                  Get l4
+              Get l1
+    Where
+      l4 =
+        TopK group_by=[#0] limit=1 monotonic=false
+          Project #0, #1
+            Join
+              Get l2
+              Filter (#1) IS NOT NULL
+                Get materialize.public.mv
+      l3 =
+        TopK group_by=[#0] limit=1 monotonic=false
+          Project #0, #1
+            Join
+              Get l2
+              Filter (#1) IS NOT NULL
+                Get materialize.public.v
+      l2 =
+        ArrangeBy keys=[[#0]]
+          Get l1
+      l1 =
+        Reduce group_by=[#0] monotonic=false
+          Get l0
+      l0 =
+        Project #1
+          Get materialize.public.t
+
+Source materialize.public.t
+  Project #1
+Source materialize.public.mv
+  Project #0, #1
+  Filter (#1) IS NOT NULL
+
+used_indexes:  - materialize.public.v_primary_idx
+EOF
+
+# Test CrossJoin derived from a comma join without a predicate.
+query T multiline
+EXPLAIN OPTIMIZED PLAN AS TEXT FOR
+SELECT t1.a, t2.a FROM t as t1, t as t2
+----
+Explained Query
+  Let
+    CrossJoin
+      ArrangeBy keys=[[]]
+        Get l0
+      Get l0
+    Where
+      l0 =
+        Project #0
+          Get materialize.public.t
+
+Source materialize.public.t
+  Project #0
+
+EOF
+
+# Test CrossJoin derived from an INNER JOIN with a trivial ON clause.
+query T multiline
+EXPLAIN OPTIMIZED PLAN AS TEXT FOR
+SELECT t1.a, t2.a FROM t as t1 INNER JOIN t as t2 ON true
+----
+Explained Query
+  Let
+    CrossJoin
+      ArrangeBy keys=[[]]
+        Get l0
+      Get l0
+    Where
+      l0 =
+        Project #0
+          Get materialize.public.t
+
+Source materialize.public.t
+  Project #0
+
+EOF
+
+# Test InnerJoin (comma syntax).
+query T multiline
+EXPLAIN OPTIMIZED PLAN AS TEXT FOR
+SELECT t1.a, t2.a
 FROM
-  T as X
-WHERE
-  NOT EXISTS (SELECT * FROM T as Y WHERE X.a = Y.b)
-LIMIT 10
+  t as t1,
+  t as t2,
+  t as t3
+WHERE t1.b = t2.b AND t2.b = t3.b
 ----
-Source materialize.public.t (u1):
-| Project (#0, #1)
+Explained Query
+  Let
+    Project #0, #2
+      Join
+        Get l1
+        Get l1
+        Project #1
+          Get l0
+    Where
+      l1 =
+        ArrangeBy keys=[[#1]]
+          Get l0
+      l0 =
+        Filter (#1) IS NOT NULL
+          Get materialize.public.t
 
-Query:
-%0 = Let l0 =
-| Get materialize.public.t (u1)
-| Project (#0)
-| Distinct group=(#0)
+Source materialize.public.t
+  Project #0, #1
+  Filter (#1) IS NOT NULL
```

Comment:
I checked the docs and I am slightly confused, as they are not very prescriptive.

```rust
pub struct LinearOperator {
    /// Rows that do not pass all predicates may be discarded.
    #[proptest(strategy = "proptest::collection::vec(any::<MirScalarExpr>(), 0..2)")]
    pub predicates: Vec<MirScalarExpr>,
    /// Columns not present in `projection` may be replaced with
    /// default values.
    pub projection: Vec<usize>,
}
```

I think the use of "may" as opposed to "should" reflects the fact that this can be optionally implemented by the `Source` implementation. Because of that, the predicates and the projection are always repeated in the plan, and consequently the output schema is not allowed to change. If we knew statically which types of sources implement projection and filter pushdown, we can avoid duplication and of those operators in the actual plan, but we are not there yet.

I wonder whether `project` should not be called (or at least rendered as) `demand` or something similar until we can implement this differently at the source level.

Reply:
I approve of "project" being changed to "demand".

Reply:
Done.